### PR TITLE
adding multiobjective optimization in nevergrad

### PR DESCRIPTION
**Pull request by teytaud**
The doc of the main function says it all: this provides a simple tool for converting a multiobjective problem into a single objective one.

def multiobjective_minimization(functions, bad_values) -> InstrumentedFunction:
    """Given several functions, and threshold on their values (above which solutions are pointless),
    this function returns a single-objective function, correctly instrumented, the minimization of which
    yields a solution to the original multiobjective problem.
    
    functions: objective functions, to be minimized, of the original multiobjective problem.
    bad_values: bad_values[i] is a threshold above which x is pointless if functions[i](x) > bad_values[i].
    
    Returns an objective function to be minimized (it is a single objective function).
    Warning: this function is not stationary.
    The minimum value obtained for this objective function is -h,
    where h is the hypervolume of the Pareto front obtained, given bad_values as a reference point.
    When called with no argument, this monoobjective function returns an approximate Pareto front,
    reaching the current hypervolume.
    """

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

People at Dagstuhl's seminar on the Theory of Randomized Search Heuristics told me that this is a missing feature so I do it.

## How Has This Been Tested (if it applies)

I did not even check that what I do makes any sense.
It's very likely I am destroying the type testing.

## Checklist

- [x ] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -87,6 +87,21 @@ print(list(sorted(ng.optimizers.registry.keys())))
 
 The [optimization documentation](docs/optimization.md) contains more information on how to use several workers, take full control of the optimization through the `ask` and `tell` interface and some pieces of advice on how to choose the proper optimizer for your problem.
 
+## Multiobjective minimization with Nevergrad
+
+Let us minimize x1 and x2 (two objective functions) assuming that values above 2.5 are of no interest.
+```python
+import nevergrad as ng
+
+f = ng.functions.multiobjective_minimization([lambda x: x[0], lambda x: x[1]], bad_values=[2.5, 2.5])
```
Comment (jrapin) [inline README.md]: if `f` is properly instrumented you should do `ng.optimizers.CMA(instrumentation=f.instrumentation, budget=100)`

File: nevergrad/functions/multiobjective.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        """
+        for i in range(index):
+            node.prev[i].next[i] = node
+            node.next[i].prev[i] = node
+            if bounds[i] > node.cargo[i]:
+                bounds[i] = node.cargo[i]
+            
+__all__ = ["hypervolume_kmax", "hypervolume"]
+
+if __name__ == "__main__":
```
Comment (jrapin) [inline nevergrad/functions/multiobjective.py]: this is not used, right?

File: nevergrad/functions/multiobjective.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+        nodes of the node that is reinserted are in the list.
+
+        """
+        for i in range(index):
+            node.prev[i].next[i] = node
+            node.next[i].prev[i] = node
+            if bounds[i] > node.cargo[i]:
+                bounds[i] = node.cargo[i]
+            
+__all__ = ["hypervolume_kmax", "hypervolume"]
```
Comment (jrapin) [inline nevergrad/functions/multiobjective.py]: not used

File: nevergrad/functions/multiobjective.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+    reference point *ref*.
+    """
+    if ref is None:
+        ref = numpy.array([3000000] * len(pointset[0]))
+    warnings.warn("Falling back to the python version of hypervolume "
+        "module. Expect this to be very slow.", RuntimeWarning)
+    hv = _HyperVolume(ref)
+    return hv.compute(pointset)
+
+def multiobjective_minimization(functions, bad_values) --> InstrumentedFunction:
```
Comment (jrapin) [inline nevergrad/functions/multiobjective.py]: I don't understand what are `bad_values`
When I understand all this better I can restructure and make it more compatible with the rest of the code.

File: nevergrad/functions/multiobjective.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/functions/multiobjective.py]: it's just a test --- I remove.

File: README.md (hunk 1 above)
Comment (teytaud) [reply README.md]: Ok, done!

File: nevergrad/functions/multiobjective.py (hunk 4 above)
Comment (teytaud) [reply nevergrad/functions/multiobjective.py]: bad_values[i] is the threshold at which the objective function functions[i] becomes so bad that we don't want to keep any point, even if they are good for other objective function values.

File: nevergrad/benchmark/test_experiments.py [hunk 5]
```diff
@@ -21,7 +21,7 @@ def test_experiments_registry(name: str, maker: Callable[[], Iterator[experiment
     with patch("shutil.which", return_value="here"):  # do not check for missing packages
         with datasets.mocked_data():  # mock mlda data that should be downloaded
             check_maker(maker)  # this is to extract the function for reuse if other external packages need it
-        if "mlda" not in name:
+        if "mlda" not in name and "moo" not in name:
```
Comment (teytaud) [inline nevergrad/benchmark/test_experiments.py]: I have no idea what this test is doing and why moo does not pass it.

File: README.md [hunk 6]
```diff
@@ -87,6 +87,21 @@ print(list(sorted(ng.optimizers.registry.keys())))
...
+optimizer = ng.optimizers.CMA(instrumentation=f.instrumentation, budget=100)  # 2 is the dimension, 100 is the budget.
+recommendation = optimizer.optimize(f)
+
+
+# The function embeds its Pareto-front:
+print("My Pareto front:", f())
```
Comment (jrapin) [inline README.md]: given that the calls to `f` can be performed in another process by an executor, this actually won't always work :s we'll need to discuss this :s

File: nevergrad/benchmark/test_experiments.py (hunk 5 above)
Comment (jrapin) [reply nevergrad/benchmark/test_experiments.py]: This test checks that running twice the same experiments with the same seed provide the same results. We'll investigate later why this occurs if it still does.

**Comment by jrapin** [comment]
Thinking about it, I think the simplest path is to add wrapper function:

```python
class AggregateFunction:

      def __self__(multi_objective_function: Callable[..., Tuple[float, ...]], higher_bounds: Tuple[float, ...]) -> None:
          self.multi_objective_function = multi_objective_function
          ...

      def compute_aggregate_loss(losses: Tuple[float, ...], *args, **kwargs, candidate: Candidate) -> None:
          ...

      def __call__(self, *args, **kwargs):
          losses = self.compute_aggregate_loss(*args, **kwargs)
          return self.compute_aggregate_loss(losses, *args, **kwargs)

     @property
     def pareto_front(self) -> List[Tuple[Tuple[Any, ...], Dict[str, Any]]]:
          return ...  # list of args and kwargs?
```

This way, we could add some simple hack in `minimize` to only call `multi_objective_function` remotely and compute the aggregate locally. It's not very beautiful, but should be straightforward and effective. Wdyt?

**Pull request closed without merging**